### PR TITLE
 GH-3414: Add parseJson to VariantBuilder for JSON-to-Variant conversion

### LINKED ISSUE #3414: parseJson implementation in VariantBuilder
### Describe the enhancement requested

Every consumer of parquet-variant currently has to implement their own
  JSON-to-Variant parsing (Spark, Kafka Connect, Flink, etc.). This adds
  native parseJson() static methods to VariantBuilder using Jackson
  streaming parser, ported from Apache Spark's production-proven logic.
  - Single-pass streaming (no intermediate JsonNode tree)
  - Smallest integer type selection (byte/short/int/long)
  - Decimal-first float handling, double fallback for scientific notation
  - Overloads for String input and pre-positioned JsonParser

https://github.com/apache/spark/tree/master/common/variant

### Component(s)

Core

### PR DESCRIPTION
**Pull request by gaurav7261**
### Rationale for this change
  Every consumer of `parquet-variant` currently has to independently implement JSON-to-Variant parsing. Apache Spark has one in its `common/variant` module 
  ([source](https://github.com/apache/spark/tree/master/common/variant)), our Kafka Connect S3 sink connector had to write one, and any other project (Flink, Trino, DuckDB-Java, etc.) would need to do the 
  same. Since `VariantBuilder` already provides all the low-level `append*()` primitives, `parseJson()` is the natural completion of that API — a canonical, reusable entry point for the most common use 
  case: converting a JSON string into a Variant.
  ### What changes are included in this PR?
  - **`parquet-variant/pom.xml`**: Added `jackson-core` (compile) and `parquet-jackson` (runtime) dependencies, following the same pattern as `parquet-hadoop`.
  - **`VariantBuilder.java`**: Added two public static methods:
    - `parseJson(String json)` — convenience method that creates a Jackson streaming parser internally.
    - `parseJson(JsonParser parser)` — for callers who already have a positioned parser (e.g., reading from a stream).
    - Internal helpers ported from Apache Spark's production `VariantBuilder.buildJson`:
      - `buildJson()` — recursive single-pass streaming parser handling OBJECT, ARRAY, STRING, NUMBER_INT, NUMBER_FLOAT, TRUE, FALSE, NULL.
      - `appendSmallestLong()` — selects the smallest integer type (BYTE/SHORT/INT/LONG) based on value range.
      - `tryAppendDecimal()` — decimal-first encoding for floating-point numbers; falls back to double only for scientific notation or values exceeding DECIMAL16 precision (38 digits).
  - **`TestVariantParseJson.java`**: 32 new tests covering all primitive types, objects (empty, simple, nested, null values, sorted keys, duplicate keys), arrays (empty, simple, nested, mixed types), and 
  edge cases (unicode, escaped strings, deeply nested documents, scientific notation, integer overflow to decimal, malformed JSON).
  ### Are these changes tested?
  Yes. 32 new tests in `TestVariantParseJson`.
  ### Are there any user-facing changes?
  Yes. Two new public static methods on `VariantBuilder`:
  - `VariantBuilder.parseJson(String json)` — returns a `Variant`
  - `VariantBuilder.parseJson(JsonParser parser)` — returns a `Variant`
  These are additive API additions with no breaking changes to existing APIs.
  Closes #3414

### CODE REVIEW

File: NOTICE [hunk 1]
```diff
@@ -73,3 +73,14 @@ notice:
 | See the License for the specific language governing permissions and
 | limitations under the License.
 
+--------------------------------------------------------------------------------
+
+This project includes code from Apache Spark with the following copyright
+notice:
+
+  Apache Spark
```
Comment (steveloughran) [inline NOTICE]: do cross-ASF projects need this credit? What I do think is good is ensure the original authors get credit in the final commit message

File: parquet-variant/src/test/java/org/apache/parquet/variant/TestVariantParseJson.java [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+package org.apache.parquet.variant;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+import java.io.IOException;
+import java.math.BigDecimal;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestVariantParseJson {
```
Comment (steveloughran) [inline parquet-variant/src/test/java/org/apache/parquet/variant/TestVariantParseJson.java:28]: 1. nested object parsing?
2. what about invalid json?
* empty file
* not a json file
* incomplete
* large json with many values

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantBuilder.java [hunk 3]
```diff
@@ -65,6 +73,152 @@ public VariantBuilder(Metadata metadata) {
...
+   * to double.
+   *
+   * <p>Ported from Apache Spark's {@code VariantBuilder.parseJson}.
+   *
+   * @param json the JSON string to parse
+   * @return the parsed Variant
+   * @throws IOException if the JSON is malformed or an I/O error occurs
+   */
+  public static Variant parseJson(String json) throws IOException {
+    try (JsonParser parser = JSON_FACTORY.createParser(json)) {
```
Comment (steveloughran) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantBuilder.java]: can be quite slow

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantBuilder.java [hunk 4]
```diff
@@ -65,6 +73,152 @@ public VariantBuilder(Metadata metadata) {
...
+   * prefer decimal encoding (no scientific notation) and fall back
+   * to double.
+   *
+   * <p>Ported from Apache Spark's {@code VariantBuilder.parseJson}.
+   *
+   * @param json the JSON string to parse
+   * @return the parsed Variant
+   * @throws IOException if the JSON is malformed or an I/O error occurs
+   */
+  public static Variant parseJson(String json) throws IOException {
```
Comment (steveloughran) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantBuilder.java]: I was to suggest making CharSequence for feeding in from other places (string fields within avro, ...) but it looks like jackson 2 doesn't support that itself.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantBuilder.java [hunk 5]
```diff
@@ -30,6 +36,8 @@
  */
 public class VariantBuilder {
 
+  private static final JsonFactory JSON_FACTORY = new JsonFactory();
```
Comment (steveloughran) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantBuilder.java]: factory should be built with some constraints so that giving it a malicious json file should be rejected rather than trigger OOM problems, etc. From the javadocs.

```
JsonFactory f = JsonFactory.builder()
    .streamReadConstraints(
        StreamReadConstraints.builder()
            .maxNestingDepth(500)
            .maxStringLength(10_000_000)
            .maxDocumentLength(5_000_000)
            .build()
    )
    .build();
```

**Review by steveloughran — COMMENTED** [review]
I think the json parsing should be in a class alongside VariantBuilder, rather than inside it, some `VariantJsonParser` class.

1. ensures that use of jackson classes are isolated; the core variant does not need jackson on the CP
2. if there is anything which can be done to improve single line JSON content performance, then it can be done there as it's life will span the whole file, rather than a row

Once it's split out, a new question surfaces: should it belong in parquet-jackson? Personally, I think it should

File: parquet-variant/pom.xml [hunk 6]
```diff
@@ -46,6 +46,17 @@
       <artifactId>parquet-column</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>${jackson.groupId}</groupId>
```
Comment (steveloughran) [inline parquet-variant/pom.xml:50]: comes with the parquet-jackson module; no need to reimport at a different scope

**Review by steveloughran — COMMENTED** [review]
Commented. As I said before, I think this should go into parquet-jackson.

I see you've added jackson-core to the classpath. Is that because the parquet-jackson ones are shaded? If the class was in that module then it'd not be an issue.

of course, adding parquet-variant to the parquet-jackson module dependency list is its own complications, but I don't see any loops in the build.

Why not raise this PR on the dev list and bring up the "where should it go?" question.

File: parquet-variant/pom.xml (hunk 6 above)
Comment (gaurav7261) [reply parquet-variant/pom.xml:50]: Hi @steveloughran , email is sent to `dev@parquet.apache.org`, can you please check there

Comment (gaurav7261) [reply parquet-variant/pom.xml:50]: Hi @steveloughran , saw your reply on email, so all looks good to you in above PR

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        break;
+      case VALUE_NULL:
+        builder.appendNull();
+        break;
+      default:
+        throw new JsonParseException(parser, "Unexpected token " + token);
+    }
+  }
+
+  private static void buildJsonObject(VariantBuilder builder, JsonParser parser) throws IOException {
```
Comment (steveloughran) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java:132]: nit: add javadoc explaining how this will co-recurse into buildJSON

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    VariantObjectBuilder obj = builder.startObject();
+    while (parser.nextToken() != JsonToken.END_OBJECT) {
+      obj.appendKey(parser.currentName());
+      parser.nextToken();
+      buildJson(obj, parser);
+    }
+    builder.endObject();
+  }
+
+  private static void buildJsonArray(VariantBuilder builder, JsonParser parser) throws IOException {
```
Comment (steveloughran) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java:150]: same comment as above: mention co-recursion

**Review by steveloughran — APPROVED** [review]
I"m happy. Reviewed code and suggested some javadocs. Co-recursion is rare in production procedural code, so worth highlighting.

Test coverage good

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java (hunk 8 above)
Comment (gaurav7261) [reply parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java:150]: done @steveloughran , please check

**Comment by steveloughran** [comment]
gaurav, you'll need to ask on the mail list for approval to run the workflows. Best to only do incremental changes on a PR

**Comment by gaurav7261** [comment]
@steveloughran i don't have access to email archive, @alamb  created thread for me, @alamb  can you please help here?

**Comment by gaurav7261** [comment]
@steveloughran can you please help here

**Comment by gaurav7261** [comment]
@wgtmac i have fixed mvn spotless, please rerun

**Comment by gaurav7261** [comment]
@wgtmac java version issue in local, now fixed with java 17 spotless,please rerun again, thanks

**Comment by gaurav7261** [comment]
@wgtmac can you please help run workflow

**Comment by gaurav7261** [comment]
@emkornfield  CI passed, can we merge now please

**Comment by emkornfield** [comment]
@gaurav7261 I think we need a committer to take a look.  CC @wgtmac @Fokko @julienledem if you have time, otherwise I can take a look but I'm not super familiar with the java side of things.

File: .gitignore [hunk 9]
```diff
@@ -20,4 +20,5 @@ target/
 mvn_install.log
 .vscode/*
 .DS_Store
+.memsearch/
```
Comment (wgtmac) [inline .gitignore:23]: Do we really need it in this PR?

File: parquet-variant/pom.xml [hunk 10]
```diff
@@ -46,6 +46,17 @@
...
+      <artifactId>jackson-core</artifactId>
```
Comment (wgtmac) [inline parquet-variant/pom.xml:51]: Does this dependency change look good to you? @gszadovszky @Fokko

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+   * Recursively builds a Variant value from the current position of a
+   * Jackson streaming parser. Handles objects, arrays, strings, numbers
+   * (int/long/decimal/double), booleans, and null.
+   */
+  private static void buildJson(VariantBuilder builder, JsonParser parser) throws IOException {
+    JsonToken token = parser.currentToken();
+    if (token == null) {
+      throw new JsonParseException(parser, "Unexpected null token");
+    }
+    switch (token) {
```
Comment (wgtmac) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java:94]: Is there any chance to parse types like date and timestamp?

**Review by wgtmac — APPROVED** [review]
+1 from my side

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java (hunk 11 above)
Comment (gaurav7261) [reply parquet-variant/src/main/java/org/apache/parquet/variant/VariantJsonParser.java:94]: i think jackson will treat them string only

File: .gitignore (hunk 9 above)
Comment (gaurav7261) [reply .gitignore:23]: not as such @wgtmac , if you need i can remove it

**Comment by gaurav7261** [comment]
@wgtmac  @Fokko can you please review and merge

**Comment by wgtmac** [comment]
Thanks @gaurav7261 for working on this and @steveloughran for the review!

**Comment by gaurav7261** [comment]
thanks @wgtmac, @steveloughran  , i see release 1.17.1 is released 5 hrs ago, just want to know  by when next release will get release so that i can use same in flink parquet and confluent kafka s3 sink connector

**Comment by wgtmac** [comment]
1.17.1 is a patched release with bug fixes only. For this feature, it is targeted in the 1.18.0 which I don't have an ETA yet. Perhaps you may want to reply to the relevant discussion thread in the dev@parquet.apache.org.
